### ISSUE TITLE
RStudio (1.2.5) do not find Rtools and offer to install even if already installed

### ISSUE THREAD

**Issue by Iuiu1234** [rstudio/rstudio#5713]
Same issue as rstudio/issues/3563 happens again with new Rstudio version.

RStudio Edition : Desktop
RStudio Version : 1.2.5019
OS Version : Windows 10
R Version : 3.6.1
Rtools : 3.5

BINREF env var & path doesn't fix it

**Comment by kevinushey** [comment]
What is the output of the following script for you?

```
.rs.isRtoolsOnPath()
.rs.addRToolsToPath()
.rs.isRtoolsOnPath()

Sys.which("make.exe")

path <- tempfile("test-", fileext = ".c")
writeLines("void test() {}", con = path)

R <- file.path(R.home("bin"), "R.exe")
norm <- normalizePath(path, winslash = "/", mustWork = TRUE)
system2(R, c("CMD", "SHLIB", shQuote(norm)))
```

**Comment by mirh** [comment]
> ELI5 why function `doAddRtoolsToPathIfNecessary` bails out, even though *scanForRTools* doesn't assure a dime about it?
https://github.com/rstudio/rstudio/blob/cf5076a88a219a275a2d128191c12c1f8f4e3890/src/cpp/session/modules/build/SessionBuildEnvironment.cpp#L137-L145

> Also here, R/sh/make/whatever has problems with paths with spaces
https://github.com/rstudio/rstudio/blob/cf5076a88a219a275a2d128191c12c1f8f4e3890/src/cpp/core/r_util/RToolsInfo.cpp#L165

**Comment by kevinushey** [comment]
> ELI5 why function doAddRtoolsToPathIfNecessary bails out, even though scanForRTools doesn't assure a dime about it?

Can you elaborate? I don't quite understand what you mean.

> Also here, R/sh/make/whatever has problems with paths with spaces

I'm not sure what we can do about this. You shouldn't try to install the toolchain into a path with spaces.

**Comment by mirh** [comment]
> Can you elaborate? I don't quite understand what you mean.

Like, the function is called *do****Add****RtoolsToPathIfNecessary*...
And it is made to return false if the scan function finds something in the hardcoded positions. 

Where would the "adding" part happen at all?

> You shouldn't try to install the toolchain into a path with spaces.

I mean, was it 1995 when windows had the default folder for installations be "program files"?

**Comment by kevinushey** [comment]
>Like, the function is called doAddRtoolsToPathIfNecessary...
> And it is made to return false if the scan function finds something in the hardcoded positions.

I think I understand now -- will put together a PR.

> I mean, was it 1995 when windows had the default folder for installations be "program files"?

Right, but this is ultimately out of our hands -- R Core maintains R and the toolchain used to compile packages for R, and to the best of my knowledge, installation of that toolchain into a path containing spaces is unsupported.

RStudio's role here is to help R to find and use Rtools toolchains that may have been installed in non-default locations, but we're still limited by what R + Rtools support themselves.

**Comment by mirh** [comment]
So.. I don't remember even half the headbanging I did before coming to smell those lines in #3563 
(and if it even was all on R, I can't be bothered to check properly on a fresh system now)

Though, [as far](https://cran.r-project.org/bin/windows/base/rw-FAQ.html#How-do-I-install-R-for-Windows_003f) as [I know](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=939), there's no special spite for spaces. If they end up not working, it's supposed to be a [bug](https://www.r4photobiology.info/2018/05/r-3-5-0-and-spaces-in-file-paths/) and fixed. 
I guess it would be ironic if the guys that program in fortran were bottlenecked by you? /s

**Comment by Iuiu1234** [comment]
R fins Rtools. Is RStudio that forces to download again.

> .rs.isRtoolsOnPath()
[1] TRUE

> .rs.addRToolsToPath()
[1] FALSE

> .rs.isRtoolsOnPath()
[1] TRUE

> Sys.which("make.exe")
                        make.exe 
"C:\\Rtools\\3.5\\bin\\make.exe" 

> path <- tempfile("test-", fileext = ".c")

> writeLines("void test() {}", con = path)

> R <- file.path(R.home("bin"), "R.exe")

> norm <- normalizePath(path, winslash = "/", mustWork = TRUE)

> system2(R, c("CMD", "SHLIB", shQuote(norm)))
The system cannot find the path specified.
C:/Rtools/3.5/mingw_64/bin/gcc  -I"C:/PROGRA~1/R/R-36~1.1/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mtune=generic -c C:/Users/xavier.banos/AppData/Local/Temp/Rtmpqsvbzh/test-6681a5e1c34.c -o C:/Users/xavier.banos/AppData/Local/Temp/Rtmpqsvbzh/test-6681a5e1c34.o
C:/Rtools/3.5/mingw_64/bin/gcc -shared -s -static-libgcc -o C:/Users/xavier.banos/AppData/Local/Temp/Rtmpqsvbzh/test-6681a5e1c34.dll tmp.def C:/Users/xavier.banos/AppData/Local/Temp/Rtmpqsvbzh/test-6681a5e1c34.o -LC:/PROGRA~1/R/R-36~1.1/bin/x64 -lR

**Comment by kevinushey** [comment]
> The system cannot find the path specified.

I wonder what's causing this error to be emitted. Either way though, it seems like the sample file was compiled successfully?

I'm very surprised that this could succeed but `.Call("rs_canBuildCpp")` would not, since that's basically exactly what we do:

https://github.com/rstudio/rstudio/blob/0691592b8743cbe291317af331a0adc5526f1ca0/src/cpp/session/modules/build/SessionBuild.cpp#L2212-L2255

Can you think of anything else that could be affecting this? Do you do anything special in your startup R profile / R environ or similar?

**Comment by kevinushey** [comment]
Closing due to lack of follow-up -- please re-open if you can provide extra information to help us understand what's going on.
